core/src/main/scala/org/bitcoins/core/gcs/GCS.scala

```diff
@@ -0,0 +1,227 @@
+package org.bitcoins.core.gcs
+
+import org.bitcoins.core.number.{UInt64, UInt8}
+import org.bitcoins.core.protocol.CompactSizeUInt
+import org.bouncycastle.crypto.macs.SipHash
+import org.bouncycastle.crypto.params.KeyParameter
+import scodec.bits.{BitVector, BinStringSyntax, ByteVector}
+
+import scala.annotation.tailrec
+
+// TODO: Replace ByteVector with a type for keys
+/**
+  * Defines all functionality dealing with Golomb-Coded Sets
+  * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#GolombCoded_Sets]]
+  */
+object GCS {
+
+  /**
+    * Given parameters and data, golomb-encodes the data
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#set-construction]]
+    */
+  def buildGCS(
+      data: Vector[ByteVector],
+      key: ByteVector,
+      p: UInt8,
+      m: UInt64): BitVector = {
+    val hashedValues = hashedSetConstruct(data, key, m)
+    val sortedHashedValues = hashedValues.sortWith(_ < _)
+    encodeSortedSet(sortedHashedValues, p)
+  }
+
+  /**
+    * Given parameters and data, constructs a GolombFilter for that data
+    */
+  def buildGolombFilter(
+      data: Vector[ByteVector],
+      key: ByteVector,
+      p: UInt8,
+      m: UInt64): GolombFilter = {
+    val encodedData = buildGCS(data, key, p, m)
+
+    GolombFilter(key, m, p, CompactSizeUInt(UInt64(data.length)), encodedData)
+  }
+
+  /**
+    * Given data, constructs a GolombFilter for that data using Basic Block Filter parameters
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#block-filters]]
+    */
+  def buildBasicBlockFilter(
+      data: Vector[ByteVector],
+      key: ByteVector): GolombFilter = {
+    buildGolombFilter(data, key, BlockFilter.P, BlockFilter.M)
+  }
+
+  private def sipHash(item: ByteVector, key: ByteVector): UInt64 = {
+    // https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#hashing-data-objects
+    val sipHashCParam = 2
+    val sipHashDParam = 4
+
+    val sh = new SipHash(sipHashCParam, sipHashDParam)
+
+    val keyParam = new KeyParameter(key.toArray)
+
+    sh.init(keyParam)
+
+    val offset = 0
+
+    sh.update(item.toArray, offset, item.length.toInt)
+
+    val digest = sh.doFinal()
+
+    UInt64.fromHex(digest.toHexString)
+  }
+
+  /**
+    * Hashes the item to the range [0, f)
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#hashing-data-objects]]
+    */
+  def hashToRange(item: ByteVector, f: UInt64, key: ByteVector): UInt64 = {
+    val hash = sipHash(item, key)
+
+    val bigInt = (hash.toBigInt * f.toBigInt) >> 64
```

Comment:
@nkohen Is this shift right going to be a problem with something that has the upper most sign bit set? I think this is going to be a bug later on, do we have any test vectors where the data element starts with 

`0x8....` (sign bit set!)

Reply:
So `0x8000` is `32768` and `0x800` is `2048` and here is my repl session:
```
scala> BigInt(32768)
res0: scala.math.BigInt = 32768

scala> res0 >> 4
res1: scala.math.BigInt = 2048
```

The reason that this works is because the `>>` defined on `BigInt` respects the arbitrary-precision abstraction and doesn't take into account something like most significant byte or anything like that.